MapboxCoreNavigation/RouteController.swift

```diff
@@ -424,19 +424,23 @@ extension RouteController: CLLocationManagerDelegate {
 
     func updateRouteLegProgress(for location: CLLocation) {
         let currentDestination = self.routeProgress.currentLeg.destination
-        guard let remainingVoiceInstructions = routeProgress.currentLegProgress.currentStepProgress.remainingSpokenInstructions else { return }
 
-        if self.routeProgress.currentLegProgress.remainingSteps.count <= 1, remainingVoiceInstructions.count == 0, currentDestination != self.previousArrivalWaypoint {
-            self.previousArrivalWaypoint = currentDestination
+        guard self.routeProgress.currentLegProgress.remainingSteps.count <= 1 else { return }
 
-            self.routeProgress.currentLegProgress.userHasArrivedAtWaypoint = true
+        if let remainingVoiceInstructions = routeProgress.currentLegProgress.currentStepProgress.remainingSpokenInstructions {
```

Comment:
If I understand this right, if we have no voiceInstructions, then we will progress the legIndex on the second to last step, which might be very long.

To be honest I don't like the previous code here as well.

How about we rewrite this code to:

- Check if user is on the second to last step
- If the delegate is implemented, we ask the delegate: self.delegate?.routeController?(self, didArriveAt: currentDestination)
- If the delegate is not implemented, I would propose something like (distance remaining to next WP < 50m or duration remaining < 5s or something like this and then advance to next leg. 

Unless I am missing anything, we should also only set 

```
        self.previousArrivalWaypoint = currentDestination
        self.routeProgress.currentLegProgress.userHasArrivedAtWaypoint = true
```

If the delegate returns true.

Reply:
> If I understand this right, if we have no voiceInstructions, then we will progress the legIndex on the second to last step, which might be very long.

Yes, that's the pre-existing logic. If you're saying it's weird, I agree with you. Maybe it was never relevant before because mapbox always included voice instructions? I honestly don't know.

> If the delegate is implemented, we ask the delegate: self.delegate?.routeController?(self, didArriveAt: currentDestination)

Just in case there is confusion here, on iOS, the `did` in a method name connotes a hook called after something has already happened.

e.g. `viewDidAppear` is a hook called after the view appears, `application(_:didReceiveRemoteNotification:)` is a hook called after the application receives a remote notification. Contrast it with `will` which is called just before the thing happens. 

- `viewWillAppear`  <- user hook
-  (*view appears*)  <- system does this
- `viewDidAppear` <- user hook

re: `didArriveAt:`

>  - returns: True to advance to the next leg, if any, or false to remain on the completed leg.

So the existing boolean returned by `didArriveAt ` does not answer the question "did I arrive?". It says "you have arrived. Should I continue navigating you to the next leg?". 

I guess it's only relevant to multi-leg journeys. TBH I haven't used it.

I agree that the current logic for determining when you're considered to have have arrived is weird. I think your approach, to compare remaining distance/time, makes much more sense than what's there now. I will update that logic. 

I don't think it's worth exposing a user over-ridable delegate method to customize this behavior. Though, I don't feel strongly about that if you do. I *do* feel strongly that **iff** we create and expose such a delegate method, it should be called something like `shouldArriveAt:` (as opposed to `did...`)

Reply:
> Maybe it was never relevant before because mapbox always included voice instructions?

Yes, I think so.

> So the existing boolean returned by didArriveAt does not answer the question "did I arrive?". It says "you have arrived. Should I continue navigating you to the next leg?".

Ah ok got it, yeah I still trown off by iOS conventions, thanks.

> I don't think it's worth exposing a user over-ridable delegate method to customize this behavior

Ok, maybe we just make this whole function overwritable? As, to be honest we are thinking about overwriting this with some more specialized code in our app ;)

Reply:
I've pushed up a commit that implements this. It turns out there was already some "slop" built in. See https://github.com/maplibre/maplibre-navigation-ios/pull/58/commits/a16516771754e76b6b27e8ae688f8d9c9d3bd7d2

>> I don't think it's worth exposing a user over-ridable delegate method to customize this behavior
>
> Ok, maybe we just make this whole function overwritable? As, to be honest we are thinking about overwriting this with some more specialized code in our app ;)

Heh, I didn't think it was worthwhile because I was having a hard time imagining that someone would have a substantially different definition of what "arrival" means, but you have already proven me wrong! 😆 I'm happy to add a delegate method in that case.

Can you share what your definition of arrival would look like? I'm curious.

Reply:
We mostly have multi leg routes, so we might want to do something different for the last waypoint compared to the others, or something like that :)